2020_2021/code_acoustics/Src/triliteration.cpp

```diff
@@ -0,0 +1,98 @@
+            uint32_t time_port, uint32_t time_starboard){
+
+    uint32_t closest_time = std::min(time_port, time_starboard);
+    uint32_t farthest_time = std::max(time_port, time_starboard);
+
+    uint32_t time_diff = farthest_time - closest_time;
+    uint32_t rough_angle = TRILITERATION::estimate_rough_angle(time_diff);
+
+    uint8_t bool_starboard = (closest_time == time_starboard);
+    return std::pair<float32_t, uint8_t> (rough_angle, bool_starboard);
+}
+
+
+uint8_t TRILITERATION::estimate_longitude(uint32_t time_port, 
+            uint32_t time_starboard, uint32_t time_stern){
+
+    uint32_t time_longitude = (std::max(time_port, time_starboard) -
+            std::min(time_port, time_starboard))/2.0;
+    return (time_longitude <= time_stern);
+}
+
+
+std::pair<float32_t, float32_t> TRILITERATION::estimate_pinger_position(
+            uint32_t time_port, uint32_t time_starboard, uint32_t time_stern, 
+            float32_t intensity_port, float32_t intensity_starboard,
+            float32_t intensity_stern){
+
+    // Estimating the likely lateral/longitude for the acoustic pinger 
+    std::pair<float32_t, uint8_t> lateral_estimate = 
+            TRILITERATION::estimate_lateral(time_port, time_starboard);
+    uint8_t longitude_estimate = TRILITERATION::estimate_longitude(
+            time_port, time_starboard, time_stern);
+        
+    // Estimating the distances
+    float32_t distance_port = TRILITERATION::estimate_distance(
+            intensity_port);
+    float32_t distance_starboard = TRILITERATION::estimate_distance(
+            intensity_starboard);
+    float32_t distance_stern = TRILITERATION::estimate_distance(
+            intensity_stern);
+
+    // Averaging the distance-estimates to triliterate the 
+    // position of the pinger
+    float32_t distance_source = (distance_port + distance_starboard
+            + distance_stern) / 3.0;
+
+    // Using the likely lateral/longitude to calculate the estimated
+    // position for the acoustic pinger
+    float32_t x, y;
+    if(!longitude_estimate)
+        y = distance_source * sin(lateral_estimate.first);
+    else
+        y = distance_source * sin(-1 * lateral_estimate.first);
+    if(!lateral_estimate.second)
+        x = distance_source * cos(lateral_estimate.first) * (-1);
+    else
+        x = distance_source * cos(lateral_estimate.second);
+    return std::pair<float32_t, float32_t> (x, y);
+}
+
+
+uint8_t TRILITERATION::check_valid_signals(
+const uint32_t& time_port, const uint32_t& time_starboard, 
+const uint32_t& time_stern, const float32_t& intensity_port, 
+const float32_t& intensity_starboard, const float32_t& intensity_stern){
+/**
+ * Here it is difficult to estimate the signals, as the signals must be
+ * evaluated in both regards to the signal intensity and the signal time 
+*/
+if((float32_t)(abs(time_port - time_starboard) * DSP_CONSTANTS::SAMPLE_TIME) > TRILITERATION::maximum_time_diff 
```

Comment:
Would be nice to add a helper function to avoid the duplicates here. Something along the lines of time_diff(time_1, time_2) or time_diff_exceeds_max(time_1, time_2)